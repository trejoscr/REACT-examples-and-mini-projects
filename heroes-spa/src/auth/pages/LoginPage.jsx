import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {

  const {login} = useContext(AuthContext);
  const navigate = useNavigate();

  const onLogin = () => {
    const lastPath = localStorage.getItem('lastPath') || '/';
    login('Andres MT');

    navigate(lastPath, {
      replace: true
    })  
  };

  return (
    <>
      <div className="container py-4">
        <div className="row">
          <div className="col-12">
            <h1>Login Page</h1>
            <hr />
            <button
              onClick={onLogin}
              className="btn btn-primary"
            >
              Login
            </button>
          </div>
        </div>
      </div>
    </>
  )
}
