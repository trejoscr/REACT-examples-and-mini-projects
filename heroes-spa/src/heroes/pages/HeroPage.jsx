import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";
import { useMemo } from "react";

export const HeroPage = () => {

  const {id} = useParams();
  const navigate = useNavigate();
  const hero = useMemo( () => getHeroById(id), [id]);

  const onNavigateBack = () => {
    navigate(-1);
  };
  
  
  if(!hero){
    return <Navigate to="/marvel"/>
  }

  return (
    <div className="row">
      <div className="col-12 col-lg-4 animate__animated animate__fadeInLeft">
        <img
          src={`/heroes/${id}.jpg`}
          alt={hero.superhero}
          className="img-thumbnail"/>
      </div>
      <div className="col-12 col-lg-8 p-4 p-lg-2 animate__animated animate__fadeInRight">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"><b>Alter ego:</b> {hero.alter_ego}</li>
          <li className="list-group-item"><b>Publisher:</b> {hero.publisher}</li>
          <li className="list-group-item"><b>First appearance:</b> {hero.first_appearance}</li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button
          className="btn btn-outline-dark"
          onClick={onNavigateBack}
        >
          Back
        </button>

      </div>
    </div>
  )
}
