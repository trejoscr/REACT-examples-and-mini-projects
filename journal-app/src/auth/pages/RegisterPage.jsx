import { Button, Grid, Link, TextField, Typography, Alert } from "@mui/material"
import { Link as RouterLink } from "react-router-dom"
import { AuthLayout } from "../layout/AuthLayout"
import { useForm } from "../../hooks"
import { useState } from "react"
import { useDispatch,useSelector } from 'react-redux'
import { startCreatingUserWithEmailPassword } from "../../store/auth/thunks"
import { useMemo } from "react"

const formData = {
  email: '',
  password: '',
  displayName: ''
}

const formValidations = {
  email: [
    (value) => /^[\w-]+(\.[\w-]+)*@([\w-]+\.)+[a-zA-Z]{2,7}$/.test(value),
    'El correo debe tener un formato válido',
  ],
  password: [(value) => value.length >= 6, 'El password debe de tener mas 6 caracteres'],
  displayName: [(value) => value.length >= 1, 'El nombre es obligatorio'],
}

export const RegisterPage = () => {

  const dispatch = useDispatch();
  const [formSubmited, setFormSubmited] = useState(false);

  const {status, errorMessage} = useSelector(state => state.auth);
  const isCheckingAuthentication = useMemo( () => status === 'checking', [status]);

  const {
    formState,
    displayName,
    email,
    password,
    onInputChange,
    isFormValid,
    displayNameValid,
    emailValid,
    passwordValid,
  } = useForm(formData, formValidations);

  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmited(true);
    if (!isFormValid) return;
    dispatch(startCreatingUserWithEmailPassword(formState));
  };

  return (
    <AuthLayout title="Crear cuenta">
      <form className="animate__animated animate__fadeIn">
          <Grid container>

          <Grid item xs={12} sx={{mt:2}}>
              <TextField
                label="Nombre completo"
                type="text"
                placeholder='Nombre completo'
                fullWidth
                name="displayName"
                value={displayName}
                onChange={onInputChange}
                error={!!displayNameValid && formSubmited}
                helperText={displayNameValid}
              />
            </Grid>
            
            <Grid item xs={12} sx={{mt:2}}>
              <TextField
                label="Correo"
                type="email"
                placeholder='correo@gmail.com'
                fullWidth
                name="email"
                value={email}
                onChange={onInputChange}
                error={!!emailValid && formSubmited}
                helperText={emailValid}
              />
            </Grid>

            <Grid item xs={12} sx={{mt:2}}>
              <TextField
                label="Contraseña"
                type="password"
                placeholder='contraseña'
                fullWidth
                name="password"
                value={password}
                onChange={onInputChange}
                error={!!passwordValid && formSubmited}
                helperText={passwordValid}
              />
            </Grid>

            <Grid container spacing={2} sx={{mb:2, mt:1}}>

              <Grid item xs={12} display={ !!errorMessage ? '' : 'none' }>
                <Alert severity="error">{errorMessage}</Alert>
              </Grid>

              <Grid item xs={12}>
                <Button
                  disabled={isCheckingAuthentication}
                  variant="contained"
                  fullWidth
                  type="submit"
                  onClick={onSubmit}
                >
                  Crear Cuenta
                </Button>
              </Grid>

            </Grid>

            <Grid container direction='row' justifyContent='end'>
              <Typography sx={{mr:1}}>Ya tienes una cuenta?</Typography>
              <Link component={RouterLink} color='inherit' to="/auth/login">
                Ingresar
              </Link>              
            </Grid>
          </Grid>
        </form>
    </AuthLayout>       

  )
}