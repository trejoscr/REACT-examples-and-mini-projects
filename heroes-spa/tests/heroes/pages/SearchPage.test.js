import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, Route } from "react-router-dom"
import { SearchPage } from "../../../src/heroes/pages/SearchPage"

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en SearchPage', () => {

  beforeEach(() => jest.clearAllMocks() );

  test('Debe de mostrarse correctamente con valores por defecto correctamemte', () => {

    const {container} = render(
      <MemoryRouter>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  
  });

  test('Debe de mostrar a Batman y el input con el valor del queryString', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');

    const img = screen.getByRole('img');
    expect(img.src).toContain('/heroes/dc-batman.jpg');

    const alert = screen.getByLabelText('alert-danger');
    expect(alert.style.display).toBe('none');
  
  });

  test('Debe de mostrar un error si no se muestra el Hero', () => {

    render(
      <MemoryRouter initialEntries={['/search?q=nonexistenthero']}>
        <SearchPage></SearchPage>
      </MemoryRouter>
    );

    const errorAlert = screen.getByLabelText('alert-danger');
    expect(errorAlert.style.display).toBe('');
  });


  test('Debe de mostrar el navigate a la pantalla nueva', () => {

    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage/>
      </MemoryRouter>
    );

    const input = screen.getByRole('textbox');
    fireEvent.change(input, {target: {name:'searchText', value: 'superman'}});

    const form = screen.getByRole('form');
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith('?q=superman');

  });

})