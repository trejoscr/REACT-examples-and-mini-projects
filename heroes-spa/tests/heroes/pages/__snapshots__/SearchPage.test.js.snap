// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en SearchPage Debe de mostrarse correctamente con valores por defecto correctamemte 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-12 col-lg-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search hero ..."
          type="text"
          value=""
        />
        <button
          class="btn btn-dark mt-2"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-12 col-lg-7"
    >
      <h4>
        Result
      </h4>
      <hr />
      <div
        class="alert alert-success animate__animated animate__fadeIn"
      >
        Search Hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
